[Kailash201] iP

Kailash201 opened the pull request:
# Mr Red
> "Making your life easier, line by line." - Mr. Red


Does this sound like you?
- Tired of forgetting awesome ideas
- Wanting to organize life
- Finding a software to see all of the important events at a glance

**Fret not Mr. Red got you covered!**

Want Mr. Red now?

Do these simple steps:
1. Visit [here](https://github.com/Kailash201/ip/releases/tag/v0.1)
1. Click duke.jar under assets

duke.jar can be run in the terminal by `java -jar duke.jar`

Features:
- [x] Simple to use 🐣
- [x] Text-based 📝
- [x] handles deadlines, events and todo 📅
- [x] Search by keyword 🔎
- [ ] Search by dates (Coming Soon)
- [ ] GUI (Coming Soon)

If you are a Java expert and want to add your own command, you can download the source file too.
You simply need to add a command class to the command package to add a new command.
An example of ByeCommand.java
```java
/**
 * Represents a type of command that can be read by the chatbot.
 */
public class ByeCommand extends Command {

    /**
     * Executes the bye command.
     * @param tasks List of tasks.
     * @param dF The file to be edited on.
     */
    @Override
    public void execute(TaskList tasks, DataFile dF) {
        System.out.println(this);
    }

    /**
     * Returns true as this is an exit command.
     * @return True.
     */
    @Override
    public boolean isExit() {
        return true;
    }

    /**
     * Returns the string representation of bye command.
     * @return String representation of bye command.
     */
    @Override
    public String toString() {
        return "Bye. Hope to see you again soon!";
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Kailash201 committed:
Duke.java

Kailash201 committed:
Duke.java

Kailash201 committed:
Duke.java

Kailash201 committed:
Duke.java, Task.java

Kailash201 committed:
add new classes and edit duke

Kailash201 committed:
test files

Kailash201 committed:
add dukception

Kailash201 committed:
lvl6 and added delete to testcases

Kailash201 committed:
added enums

Kailash201 committed:
Add save function

Kailash201 committed:
Merge branch 'Level-7'

Kailash201 committed:
Add date-time function

Kailash201 committed:
Merge branch 'branch-Level-8'

Kailash201 committed:
Add more OOP classes

Kailash201 committed:
Edit oop classes

Kailash201 committed:
Merge branch 'branch-A-MoreOOP'

Kailash201 committed:
Add classes into packages

Kailash201 committed:
Merge branch 'branch-A-Packages'

Kailash201 committed:
Merge remote-tracking branch 'origin/add-gradle-support'

Kailash201 committed:
Add gradle

Kailash201 committed:
Merge branch 'branch-A-Gradle'

Kailash201 committed:
Add 1 text directory file

Kailash201 committed:
More OOP

Kailash201 committed:
Add JUnit for 3 classes

Kailash201 committed:
Merge branch 'branch-A-JUnit'

Kailash201 committed:
Add find function

Kailash201 committed:
Add Java docs for methods/classes

Kailash201 committed:
Fix coding standards

Kailash201 committed:
Merge branch 'branch-Level-9'

Kailash201 committed:
Fix merge conflict

Kailash201 committed:
Fix merge conflict

leontan2 commented on src/main/java/parser/Parser.java:
```diff
@@ -0,0 +1,204 @@
+    public Parser(){
+    }
+
+    /**
+     * Parser constructor that takes in a String.
+     * @param type The type of task.
+     */
```
Perhaps you may consider changing the phrasing to be "Constructs..." instead of the current "Constructor..."?

Kailash201 replied:
Thx for the suggestion, will take note of it

leontan2 commented on src/main/java/parser/Parser.java:
```diff
@@ -0,0 +1,204 @@
+            if (input.indexOf("/by") - 1 < 6) {
+                throw new DukeException("Your Event name is blank");
+            }
+            String desc = input.substring(6, input.indexOf("/from") - 1);
+            if (desc.isBlank() || desc.isEmpty()) {
+                throw new DukeException("Your Event name is blank");
+            }
+            texts.add(desc);
+            String from = input.substring(input.indexOf("/from") + 6, input.indexOf("/to") - 1);
+            texts.add((from));
+            String to = input.substring(input.indexOf("/to") + 4);
+            texts.add(to);
+            break;
+        }
+        case "D": {
+            if (input.indexOf("/by") - 1 < 9) {
+                throw new DukeException("Your Deadline name is blank");
+            }
+            String desc = input.substring(9, input.indexOf("/by") - 1);
+            if (desc.isBlank() || desc.isEmpty()) {
+                throw new DukeException("Your Deadline name is blank");
+            }
+            texts.add(desc);
+            String by = input.substring(input.indexOf("/by") + 4);
+            texts.add(by);
+            break;
+        }
+        case "T": {
+            String desc = input.substring(input.indexOf("todo") + 5);
+            if (desc.isBlank() || desc.isEmpty()) {
+                throw new DukeException("Your Todo name is blank");
+            }
+            texts.add(desc);
+            break;
+        }
+        case "P": {
+            String keyWord = splitter(input, " ").get(1);
+            texts.add(keyWord);
+            break;
+        }
```
I like the good use of switch statements here, as well as its proper formatting and indentation

leontan2 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,66 @@
+import tasks.Task;
+import tasks.TaskList;
+import ui.Ui;
+import java.io.*;
```
Should wildcard imports be used? Maybe it is better to list them out?

Kailash201 replied:
yup, thx for pointing it out

leontan2 commented on src/main/java/storage/DataFile.java:
```diff
@@ -0,0 +1,175 @@
+                    .append(System.lineSeparator());
+        } else if (task instanceof Deadline) {
+            Deadline dl = (Deadline) task;
+            custom.append("D").append(div).append(dl.getDesc()).append(div).append(dl.getBy()).append(div)
+                    .append(dl.getStatus().equals("X") ? 1 : 0)
+                    .append(System.lineSeparator());
```
Good implementation of separation of lines when the code is too long.

leontan2 commented on src/main/java/tasks/TaskList.java:
```diff
@@ -0,0 +1,61 @@
+ */
+public class TaskList {
+
+    ArrayList<Task> tasks;
```
Perhaps you can include a visibility marker here.

Kailash201 replied:
Sorry, what you mean by that?

leontan2 commented on src/main/java/commands/FindCommand.java:
```diff
@@ -0,0 +1,27 @@
+public class FindCommand extends Command{
+    private final String keyWord;
+
+    public FindCommand(String keyWord) {
```
Would it help with clarity if perhaps you can include JavaDoc comments for this FindCommand file?

leontan2 commented on src/test/java/parser/ParserTest.java:
```diff
@@ -0,0 +1,88 @@
+
+public class ParserTest {
+    @Test
+    public void convertTaskInput_deadlineCommand_success() throws DukeException {
```
I like the adherence to the suggested structure "featureUnderTest_testScenario_expectedBehavior()" as per the coding standard for test cases.

leontan2 reviewed (commented):
Overall, coding standard expectations were well met. Everything was well documented and adhered to. The main comments were with regards to some missing JavaDoc comments and the use of wildcard imports.

Kailash201 committed:
Add GUI and Checkstyle

Kailash201 committed:
Merge branch 'branch-Level-10'

Kailash201 committed:
Add assert to some classes and clean up code

Classes could possible have potential bugs.

Let's add asserts to Parser, DataFile and some in Command classes as
it contains the most of the logic. For example, checking if
storage file exists when it is known that the storage file must
exists.

Kailash201 committed:
Merge pull request #1 from Kailash201/branch-A-Assertions

Add assertions

Kailash201 committed:
Merge branch 'master' into branch-A-CodeQuality
Merge master to feature brach.

Kailash201 committed:
Improve code quality

They are some classes where code can be improved.

Refactor codes in some classes. Like for an example,
in MainWindow.java, use extract method for one of the methods
in it.

Kailash201 committed:
Refactor Parser and Datafile class

Parser and Datafile classes had long methods such as getCommand
in Parser.java.

Use extract method for methods in parser and datafile classes
to eliminate long methods.

Kailash201 committed:
Merge pull request #2 from Kailash201/branch-A-CodeQuality

Improved code

Kailash201 committed:
Add FixedDuration task feature.

Sometimes user wants to set the duration
of a task so that they can keep track.

Add new command called todotime to allow users
to add a task and also describes how long it takes.

Kailash201 committed:
Fix TaskDuration bug

Kailash201 committed:
Merge branch 'branch-B-FixedDurationTasks'

Kailash201 committed:
Improve code

Kailash201 committed:
Fix out of index error bugs

Kailash201 committed:
Add help feature

Kailash201 committed:
Merge branch 'branch-Help'

Kailash201 committed:
Edit chatbot theme

Kailash201 committed:
Update README.md file

Kailash201 committed:
Update README.md

Kailash201 committed:
Update README.md

Kailash201 committed:
Fix minor bugs

Kailash201 committed:
Fix minor bugs

Kailash201 committed:
Minor changes on FXMl

Kailash201 committed:
Add more unit test

Kailash201 committed:
Make minor changes

Kailash201 committed:
Update README.md

Kailash201 committed:
Add launcher class to fix windows bug

Kailash201 committed:
Update screen size

Kailash201 committed:
Update GUI

Kailash201 committed:
update ss
